src/extra/widgets/animbtn/lv_animbtn.c

```diff
@@ -0,0 +1,309 @@
+/**
+ * @file lv_animbtn.c
+ *
+ */
+
+/*********************
+ *      INCLUDES
+ *********************/
+
+#include "lv_animbtn.h"
+
+#if LV_USE_RLOTTIE != 0 && LV_USE_ANIMBTN != 0
+
+/*********************
+ *      DEFINES
+ *********************/
+#define MY_CLASS &lv_animbtn_class
+
+/**********************
+ *      TYPEDEFS
+ **********************/
+
+/**********************
+ *  STATIC PROTOTYPES
+ **********************/
+static void lv_animbtn_constructor(const lv_obj_class_t * class_p, lv_obj_t * obj);
+static void lv_animbtn_event(const lv_obj_class_t * class_p, lv_event_t * e);
+static void apply_state(lv_obj_t * animbtn);
+static void loop_state(lv_obj_t * animbtn);
+static lv_animbtn_state_t suggest_state(lv_obj_t * animbtn, lv_animbtn_state_t state);
+static lv_animbtn_state_t get_state(const lv_obj_t * animbtn);
+static int is_state_valid(const lv_animbtn_state_desc_t * state);
+
+/**********************
+ *  STATIC VARIABLES
+ **********************/
+const lv_obj_class_t lv_animbtn_class = {
+    .base_class = &lv_obj_class,
+    .instance_size = sizeof(lv_animbtn_t),
+    .constructor_cb = lv_animbtn_constructor,
+    .event_cb = lv_animbtn_event,
+};
+
+/**********************
+ *      MACROS
+ **********************/
+
+/**********************
+ *   GLOBAL FUNCTIONS
+ **********************/
+
+/**
+ * Create an image button object
+ * @param parent pointer to an object, it will be the parent of the new image button
+ * @return pointer to the created image button
+ */
+lv_obj_t * lv_animbtn_create(lv_obj_t * parent, lv_obj_t * anim)
+{
+    LV_LOG_INFO("begin");
+    lv_obj_t * obj = lv_obj_class_create_obj(MY_CLASS, parent);
+    lv_obj_class_init_obj(obj);
+    /*Capture the animation picture*/
+    ((lv_animbtn_t *)obj)->lottie = anim;
+    lv_obj_set_parent(anim, obj);
```

Comment:
Why can't we use the rlottie object directly? I.e. why is it required to wrap it into a lv_obj?

Without it, we could you same API as lv_rlottie_create does.

Just thinking aloud: if it worked maybe we could even move all these into `lv_rlottie`. Hm... But probably keeping basic animation and the button-like functionalities separate is a better idea.

Reply:
This button is just that, a wrapper that handles event and sets the the animation state. It can be done from outside, but it's troublesome to get right, hence the object. I'm using a object here because I have to capture the PRESS/CLICKED events and match them to button state, which would not make sense for an image IMHO. Also, I'm capturing the animation's EVENT_READY event here to restart the loop upon termination, and I haven't found a way to do that without becoming a parent of it.